test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package app does all of the work necessary to configure and run a
+// Kubernetes app process.
+package app
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	cdiapi "github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	specs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"k8s.io/klog/v2"
+	drapbv1 "k8s.io/kubernetes/pkg/kubelet/apis/dra/v1alpha1"
+	plugin "k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin"
+)
+
+type pluginConfig struct {
+	driverName             string
+	draAddress             string
+	pluginRegistrationPath string
+}
+
+type examplePlugin struct {
+	config pluginConfig
+}
+
+// newExamplePlugin returns an initialized examplePlugin instance
+func newExampleDriver(config pluginConfig) (*examplePlugin, error) {
+
+	if err := os.MkdirAll(filepath.Dir(config.draAddress), 0750); err != nil {
+		return nil, err
+	}
+
+	return (&examplePlugin{
+		config: config,
+	}), nil
+}
+
+func startPlugin(config *pluginConfig) error {
+	// Run a registration server and registers plugin with kubelet
+	err := plugin.StartRegistrar(config.driverName, config.draAddress, config.pluginRegistrationPath)
+	if err != nil {
+		return err
+	}
+
+	driver, err := newExampleDriver(*config)
+	if err != nil {
+		return err
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	plugin.StartNonblockingGrpcServer(config.draAddress, driver)
+	return nil
+}
+
+var cdiVersion = "0.2.0"
+var kind = "vendor.com/device"
```

Comment:
I agree with @pohly here. 'device' is too generic name. vendor.com is ok.
I'd suggest two things regarding CDI file naming:
- get CDI directory from the command line with /var/run/cdi as a default
- use this naming convention: ```<cdi directory>/<vendor>-<device>-<claim uid or something unique>.cdi```

Reply:
We discussed this on Slack.

Let's use `<cdi directory>/<driver name>-<claim uid>.json`, with `<cdi directory` being configurable and defaulting to `/var/run/cdi`. Inside the file, let's use `k8s.io` as vendor and `<driver name>/exampledevice` as kind.

The kind doesn't need to be unique, does it?

Reply:
it doesn't, but fully qualified device name must be unique. Otherwise runtime wouldn't be able to get CDI file by device fqdn. This is not the case in the current implementation and needs to be fixed.

Reply:
@bart0sh @pohly Did you see my comment above?
When I deployed 'shared-resource-claim', uid is not unique, but name is unique.
Umm, is it

Reply:
The claim name is not unique. The same name may appear in different namespaces.

Reply:
@pohly
If <driver name>/exampleevcies is used as kind, then what  is 'vendor' for?

Reply:
@pohly And, I have no idea which slack thread you mean..

Reply:
Sorry, I mixed it up. I meant our internal Teams chat.

I don't know how CDI uses "vendor". I probably would use the driver name also as "vendor".

Reply:
> If /exampleevcies is used as kind, then what is 'vendor' for?

Can you elaborate a little bit more? Vendor is usually a domain, describing a company that develops and maintains a plugin, e.g. nvidia.com, intel.com etc. In our case we decided to use k8.io as a kind. Here is a link for the vendor description in the CDI schema file: https://github.com/container-orchestrated-devices/container-device-interface/blob/master/schema/schema.json#L11

Reply:
@bart0sh Yes, I know that `kind` used to be vendor.com/device, which I hard coded.
But, @pohly said `<driver name>/example` device as `kind`.
So, before @pohly said he would use the driver name also as `vendor`, I think it is not possible for me to understand fully where I put `vendor`.

So, `vendor`=`drivername`=`k8s.io`, am I right?

Reply:
vendor is "k8s.io", driverName is "example-driver.cdi." + vendor, kind is vendor+"/exampledevice" in my understanding

Reply:
and back to the original topic: CDI file path is <cdi directory>/<driver name>-<claim uid>-<random suffix created by os.CreateTemp>.json

Reply:
Nope. Drop the random suffix to simplify idempotency.

Reply:
@pohly what about kind? You said before `<driver name>/example-device`, but he now says `<vendor>/exmaple-device`

Reply:
@pohly @bart0sh I didn't get any answer about 'vendor'?
Please read what you were talking. Who do I need to listen to?

Reply:
we're different people, why should we talk the same thing? @pohly is a main reviewer as I've mentioned before.

Reply:
> I didn't get any answer about 'vendor'?

Here is mine in case you've missed it:
```
[bart0sh](https://github.com/bart0sh) [4 hours ago](https://github.com/pohly/kubernetes/pull/7#discussion_r941174593)
vendor is "k8s.io", driverName is "example-driver.cdi." + vendor, kind is vendor+"/exampledevice" in my understanding
```